tests/integration/ha/test_ha.py

```diff
@@ -12,72 +12,309 @@
+
+    logger.info("Waiting for Juju to detect new IP...")
+    async with ops_test.fast_forward():  # to ensure update-status runs fast enough
+        await ops_test.model.block_until(
+            lambda: leader_host
+            not in helpers.get_hosts_from_status(ops_test),  # ip changes after lxd config add
+            timeout=900,
+            wait_period=5,
+        )
+
+    logger.info("Stopping continuous_writes...")
+    cw.stop_continuous_writes()
+
+    logger.info("Counting writes on surviving units...")
+    last_write = cw.get_last_znode(
+        parent=parent, hosts=non_leader_hosts, username=USERNAME, password=password
+    )
+    total_writes = cw.count_znodes(
+        parent=parent, hosts=non_leader_hosts, username=USERNAME, password=password
+    )
+    assert last_write == total_writes
+
+    new_hosts = helpers.get_hosts_from_status(ops_test)
+    new_leader_host = max(set(new_hosts) - set(hosts))
+
+    logger.info("Checking old leader caught up...")
+    last_write_leader = cw.get_last_znode(
+        parent=parent, hosts=new_leader_host, username=USERNAME, password=password
+    )
+    total_writes_leader = cw.count_znodes(
+        parent=parent, hosts=new_leader_host, username=USERNAME, password=password
+    )
+    assert last_write == last_write_leader
+    assert total_writes == total_writes_leader
+
+
+@pytest.mark.abort_on_fail
+async def test_two_clusters_not_replicated(ops_test: OpsTest, request):
+    """Confirms that writes to one cluster are not replicated to another."""
+    zk_2 = f"{APP_NAME}2"
+
+    logger.info("Deploying second cluster...")
+    new_charm = await ops_test.build_charm(".")
+    await ops_test.model.deploy(new_charm, application_name=zk_2, num_units=3)
+    await helpers.wait_idle(ops_test, apps=[APP_NAME, zk_2])
+
+    parent = request.node.name
+
+    hosts_1 = helpers.get_hosts(ops_test)
+    hosts_2 = helpers.get_hosts(ops_test, app_name=zk_2)
+    password_1 = helpers.get_super_password(ops_test)
+    password_2 = helpers.get_super_password(ops_test, app_name=zk_2)
+
+    logger.info("Starting continuous_writes on original cluster...")
+    cw.start_continuous_writes(
+        parent=parent, hosts=hosts_1, username=USERNAME, password=password_1
+    )
+    await asyncio.sleep(30)  # letting client set up and start writing
+
+    logger.info("Checking writes are running at all...")
+    assert cw.count_znodes(parent=parent, hosts=hosts_1, username=USERNAME, password=password_1)
+
+    logger.info("Stopping continuous_writes...")
+    cw.stop_continuous_writes()
+
+    logger.info("Confirming writes on original cluster...")
+    assert cw.count_znodes(parent=parent, hosts=hosts_1, username=USERNAME, password=password_1)
+
+    logger.info("Confirming writes not replicated to new cluster...")
+    with pytest.raises(Exception):
+        cw.count_znodes(parent=parent, hosts=hosts_2, username=USERNAME, password=password_2)
```

Comment:
**todo** best to keep the assertion in the tests. It is weird that the assertion is done in the function. Using `Exception` also catches other things (like not connectinve to the cluster) which we should not catch, as that should be working. In general, I believe it makes more sense to be a bit more specific, like `assert cw.count_znodes(...) == 0`

Reply:
We want to avoid 0. As 0 == 0, we might get tests passing unexpectedly. I'll add some `NoZnodes` error and catch that specifically.
Though, this is a test, if there's ANY raise, we should fail the test, not gracefully handle it.